### PR TITLE
Assets usage cost added to transaction cost.

### LINKED ISSUE #340: Add core operations to transaction cost
Currently, transaction cost reflects only script execution cost. Core efforts to parse transaction and make needed checks are not counted. 

We can add at least assets preservation rules check to the cost function, at least. As a simplest option, we can count preservation rules implemented as a \Sigma script (but that cost could be unreasonably high).

### PR DESCRIPTION
**Pull request by oskin1**
Close #340

### CODE REVIEW

File: src/main/scala/org/ergoplatform/settings/Parameters.scala [hunk 1]
```diff
@@ -211,29 +219,38 @@ object Parameters {
   val MinValueMin = 0
   val MinValueMax = 10000 //0.00001 Erg
 
+  val TokenAccessCostDefault = 100
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/settings/Parameters.scala]: I propose to remove Min/Max values and step

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 2]
```diff
@@ -161,6 +162,29 @@ class ErgoTransactionSpec extends ErgoPropertyTest {
     }
   }
 
+  property("assets usage correctly affects transaction total cost") {
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala]: I'm not sure this test really test what expected, you take 2 different transactions and compare their costs.
I propose the following scenario:
1. Generate random transaction `tx` with tokens
2. Check, that if you add 1 token to `tx.input.head` from other inputs, `tx` cost increases by `tokenAccessCost`
3. Check, that if you add 1 token to `tx.input.head` of a new type, `tx` cost increases by `2 * tokenAccessCost`
4-5. The same 2-3 for outputs

File: src/test/scala/org/ergoplatform/mining/ErgoMinerPropSpec.scala [hunk 3]
```diff
@@ -109,17 +109,13 @@ class ErgoMinerPropSpec extends ErgoPropertyTest {
     }
 
     // transactions reach computation cost block limit
-    checkCollectTxs(100000L, Int.MaxValue, withTokens = false)
+    checkCollectTxs(100000L, Int.MaxValue)
 
     // transactions reach block size limit
-    checkCollectTxs(Long.MaxValue, 4096, withTokens = false)
-
-    // too many tokens in fees
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/mining/ErgoMinerPropSpec.scala]: test with tokens is still useful here to ensure that miner collect correct transactions from mempool even if they have tokens

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 4]
```diff
@@ -161,6 +162,40 @@ class ErgoTransactionSpec extends ErgoPropertyTest {
...
+    val txGen = validErgoTransactionGenTemplate(1, 1, 8, 16)
+    forAll(txGen) { case (from, tx) =>
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala]: Please regroup test and write comments, what exactly are you checking here. e.g.
```
// adding new token increases transaction cost by 2 TokenAccessCostDefault (one more group + one more token in total)
val txMod0 = tx.copy(inputs = tx.inputs.init :+ txInMod0)
val inputIncTxCost0 = txMod0.statefulValidity(from.init :+ modifiedIn0, emptyStateContext).get
(inputIncTxCost0 - initTxCost) shouldEqual Parameters.TokenAccessCostDefault * 2

// adding existing token increases transaction cost by TokenAccessCost 
...

```

File: src/main/resources/api/openapi.yaml [hunk 5]
```diff
@@ -611,6 +612,13 @@ components:
             blockVersion:
               $ref: '#/components/schemas/Version'
               nullable: false
+            tokenAccessCost:
+              type: integer
+              format: int32
+              description: Token access cost (in NanoErgs)
```
Comment (catena2w) [inline src/main/resources/api/openapi.yaml]: It is not in nanoErgs

File: papers/yellow/voting.tex [hunk 6]
```diff
@@ -45,6 +45,8 @@ \subsection{Parameters table}
 \hline
 4 & Maximum cumulative computational cost of a block & 1000000 & - & 16384 & - \\
 \hline
+5 & Token access cost (in nanoErgs) & 100 & - & - & - \\
+\hline
 \end{tabular}
```
Comment (catena2w) [inline papers/yellow/voting.tex]: Please also describe cost rules somewhere in the Transactions section of yellow paper

File: src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala [hunk 7]
```diff
@@ -161,6 +162,56 @@ class ErgoTransactionSpec extends ErgoPropertyTest {
...
+      // new token added to the last output
+      val modifiedOut0 = ErgoBox(out0.value, out0.proposition, out0.creationHeight,
+        out0.additionalTokens :+ additionalToken, out0.additionalRegisters, out0.transactionId, out0.index)
+      // existing token added to the last output
+      val modifiedOut1 = ErgoBox(out0.value, out0.proposition, out0.creationHeight,
+        out0.additionalTokens :+ existingToken, out0.additionalRegisters, out0.transactionId, out0.index)
+
+      // update transaction inputs and outputs accordingly
+      val txMod0 = tx.copy(inputs = tx.inputs.init :+ txInMod0) // new token group added to one input
+    val txMod1 = tx.copy(inputs = tx.inputs.init :+ txInMod1) // existing token added to one input
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/modifiers/mempool/ErgoTransactionSpec.scala]: Please format code

**Comment by kushti** [comment]
@oskin1 approved, but

1. Please take a look to https://github.com/ergoplatform/ergo/issues/645 . I guess we should have spam-simulation tests for big transaction and many tokens (like we have spam-simulating tests in the AST language interpreter), 
2. Please limit number of tokens per tx to UByte max value, at least while such a limit exist in tx serialization (and maybe such limit is enough?).

**Pull request merged**